5.next: add expect log feature to TestSuite

LordSimal opened the pull request:
Sometimes it can be quite handy to make sure something is being logged inside tests.

we do have the `ArrayEngine` which is an in memory log engine but its not really straight forward how to use it inside tests.

Therefore this PR adds the following methods to the TestSuite class:

- `$this->setupLog('debug');` which needs to be called before the Log message is actually written to the Log engine via e.g.  `$this->get('/')` or some other app specific code
- `$this->expectLogMessage('debug', 'The expected message');` which does a 1:1 comparison check between the logged message and what is being expected
- `$this->expectLogMessageContains('debug', 'expected message');` same as above but just with a `str_contains` instead of a direct comparison

As always feel free to recommend better method names 😉

LordSimal committed:
add log expect methods to TestCase

markstory commented on src/TestSuite/TestCase.php:
```diff
@@ -1105,4 +1109,76 @@ protected function expectErrorHandlerMessageMatches(string $regex, Closure $call
+     * @param array|string $engines The engine(s) which should receive a log message
+     * @return void
+     */
+    public function setupLog(array|string $engines): void
```
Should we put these methods in a trait similar to `EmailTrait` and integration testing for web and console apps?

markstory commented on src/TestSuite/TestCase.php:
```diff
@@ -1105,4 +1109,76 @@ protected function expectErrorHandlerMessageMatches(string $regex, Closure $call
+     * @param string $failMsg The error message if the message was not in the log engine
+     * @return void
+     */
+    public function expectLogMessage(string $engine, string $expectedMessage, string $failMsg = ''): void
```
Should these follow the `assertXyZ` naming conventions we use for other assertions?

markstory commented on src/TestSuite/TestCase.php:
```diff
@@ -1105,4 +1109,76 @@ protected function expectErrorHandlerMessageMatches(string $regex, Closure $call
+     * @return void
+     */
+    protected function _expectLogMessage(
+        string $engine,
```
Instead of worrying about which logger a message was captured in, should we have ways to assert the log level and scopes in addition to the message (which you have already).

LordSimal replied:
I don't know if we should go that deep into specifying log configs and expected log messages.
In tests I don't really care on which info level and/or scope a specific message is being logged as long as it is being logged (which this does)
But maybe I don't see the benefit of exactly specifying what log message is being logged in what specific context.

LordSimal replied:
just added the ability to assert log levels as well

markstory reviewed (commented):
I like this idea :clap:

markstory commented on src/TestSuite/TestCase.php:
```diff
@@ -230,6 +231,7 @@ protected function tearDown(): void
         if (class_exists(Mockery::class)) {
             Mockery::close();
         }
+        Log::reset();
```
This could be an `@after` hook for the Log trait. We use a similar pattern in the `EmailTestTrait`.

LordSimal replied:
I looked at that trait but didn't realize there is this phpdoc.

markstory commented on tests/TestCase/TestSuite/LogTraitTest.php:
```diff
@@ -0,0 +1,73 @@
+namespace Cake\Test\TestCase\TestSuite;
+
+use Cake\Log\Log;
+use Cake\TestSuite\LogTrait;
```
Isn't the trait that adds logging methods to classes also called `LogTrait`? Should we have a different class name so that it is more obvious whether the class is including methods to *do* logging or to *test* logging?

LordSimal replied:
renamed it to LogTestTrait

ADmad commented on src/TestSuite/LogTrait.php:
```diff
@@ -0,0 +1,99 @@
+    public function setupLog(array|string $engines): void
+    {
+        Log::reset();
+        if (is_array($engines)) {
```
Just cast to array and loop over the var instead of `if else`.

ADmad commented on src/TestSuite/LogTrait.php:
```diff
@@ -0,0 +1,99 @@
+        Log::reset();
+        if (is_array($engines)) {
+            foreach ($engines as $engine) {
+                Log::setConfig($engine, ['className' => 'Array']);
```
Don't we need to allow providing log configs so that one can test for specific levels or scopes for e.g.?

LordSimal committed:
rename LogTrait to LogTestTrait

markstory commented on tests/TestCase/TestSuite/LogTraitTest.php:
```diff
@@ -0,0 +1,73 @@
+        $this->setupLog(['debug', 'error']);
+        Log::debug('This usually needs to happen inside your app');
+        Log::error('Some error message');
+        $this->assertLogMessage('debug', 'This usually needs to happen inside your app');
+        $this->assertLogMessage('error', 'Some error message');
```
```suggestion
        $this->assertLogMessage('errror', 'This usually needs to happen inside your app');
        $this->assertLogMessage('debug', 'Some error message');
```

Do these assertions fail? They should, but I think based on how the loggers are configured that they will pass.

LordSimal replied:
This fails but let me add a separate test to prove that

markstory commented on src/TestSuite/LogTestTrait.php:
```diff
@@ -0,0 +1,107 @@
+        $engines = (array)$engines;
+        foreach ($engines as $engine) {
+            Log::setConfig($engine, ['className' => 'Array']);
+        }
```
Do we want to support config parameters, or apply some conventions? Like `debug` as an engine name will wire up a logger that listens for info + debug log messages, while `error` enables capture of error and greater levels.

LordSimal committed:
make sure log level messages don't get mixed up

LordSimal committed:
allow asserting log messages on specific levels

markstory commented on src/TestSuite/LogTestTrait.php:
```diff
@@ -0,0 +1,126 @@
+     * @param string $failMsg The error message if the message was not in the log engine
+     * @return void
+     */
+    public function assertLogMessage(
+        string $engine,
+        string $expectedMessage,
+        ?string $level = null,
+        string $failMsg = ''
+    ): void {
```
Is the engine a log message was recorded on the most important part of the log message? To me the goal of testing logs is to:

1. Validate that a given log message was created.
2. That the log message was emit at the right level,
3. That the log has the right scopes so that I can be confident that logs will be routed correctly.

The engine that captured the log message is less important to me than the level. What do you think about the assertion signatures being

```php
assertLogMessage(string $level, string $message, ?string $engine = null, string $failMessage = '')
```

I can put together a commit/diff for the change I have in mind.

LordSimal replied:
sure, go ahead

markstory committed:
Shift the focus of LogTestTrait to levels and scopes (#17471)

For me when I want to write assertions on logs, I'm mostly interested in
either logs of a certain level (no error logs made) or logs of a certain
scopes (security, and audit logs). What logger captures the log messages
is less important than that the application is logging.

LordSimal committed:
add missing tests

LordSimal commented:
Thanks to @markstory for adjusting this trait to be level and scope based 👍🏻
